go/vt/vtctl/workflow/server.go

```diff
@@ -850,6 +852,207 @@ func (s *Server) Materialize(ctx context.Context, ms *vtctldatapb.MaterializeSet
 	return mz.startStreams(ctx)
 }
 
+// MaterializeAddTables adds specified tables to the existing workflow.
+func (s *Server) MaterializeAddTables(ctx context.Context, req *vtctldatapb.MaterializeAddTablesRequest) error {
+	if len(req.TableSettings) == 0 {
+		return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "no tables found in the request")
+	}
+
+	targets, err := s.ts.FindAllShardsInKeyspace(ctx, req.Keyspace, nil)
+	if err != nil {
+		return err
+	}
+	targetShardInfos := maps.Values(targets)
+
+	// This will be helpful to find duplicate tables.
+	tableSet := sets.New[string]()
+	for _, ts := range req.TableSettings {
+		tableSet.Insert(ts.TargetTable)
+	}
+
+	// Store the ReadVReplicationWorkflow response for later use.
+	readVReplicationWorkflowResp := make(map[string]*tabletmanagerdatapb.ReadVReplicationWorkflowResponse)
+	var (
+		mu             sync.Mutex
+		sourceKeyspace string
+		workflowType   binlogdatapb.VReplicationWorkflowType
+	)
+
+	// Validation for duplicate reference tables.
+	err = forAllShards(targetShardInfos, func(si *topo.ShardInfo) error {
+		tablet, err := s.ts.GetTablet(ctx, si.PrimaryAlias)
+		if err != nil {
+			return err
+		}
+		res, err := s.tmc.ReadVReplicationWorkflow(ctx, tablet.Tablet, &tabletmanagerdatapb.ReadVReplicationWorkflowRequest{
+			Workflow: req.Workflow,
+		})
+		if err != nil {
+			return vterrors.Wrapf(err, "failed to read workflow %s on shard %s/%s", req.Workflow, req.Keyspace, tablet.Shard)
+		}
+
+		mu.Lock()
+		if len(res.Streams) > 0 && sourceKeyspace == "" {
+			sourceKeyspace = res.Streams[0].Bls.Keyspace
+		}
+		workflowType = res.WorkflowType
+		readVReplicationWorkflowResp[tablet.Shard] = res
+		mu.Unlock()
```

Comment:
Nit, but IMO it's worth using a closure here to be sure the mutex is unlocked using a defer (e.g. you get a panic between the lock and unlock, and that panic is recovered up the call stack, but then we don't unlock the mutex).
```
		func() {
			mu.Lock()
			defer mu.Unlock()
			if len(res.Streams) > 0 && sourceKeyspace == "" {
				sourceKeyspace = res.Streams[0].Bls.Keyspace
			}
			workflowType = res.WorkflowType
			readVReplicationWorkflowResp[tablet.Shard] = res
		}()
```

Reply:
yeah, right! done.